Wrap all sodium/nacl functions in crypto_core.c.

iphydf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/343)
<!-- Reviewable:end -->

Diadlo commented:
Reviewed 1 of 45 files at r1, 1 of 3 files at r2.
Review status: 2 of 47 files reviewed at latest revision, 7 unresolved discussions.

---

*[toxcore/crypto_core.c, line 48 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZxLmyJE974Z508nB:-KZNZxLnnAWlptpLPrW8:b-mqel8s) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L48)):*
> ```C
> #include <randombytes.h>
> #define crypto_box_MACBYTES (crypto_box_ZEROBYTES - crypto_box_BOXZEROBYTES)
> /* I know */
> ```

What do you know?)

---

*[toxcore/crypto_core.c, line 88 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNbB1NaPmQjP7-bAOF:-KZNbB1Omk4ZiVhBm5RP:b-fgb4dg) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L88)):*
> ```C
>    returns 0 if both mem locations of length are equal,
>    return -1 if they are not. */
> int public_key_cmp(const uint8_t *pk1, const uint8_t *pk2)
> ```

In `.h` file you specified `int32_t`. Same for all other `int` occupation

---

*[toxcore/crypto_core.c, line 90 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZeiW6tZKWR0Enn3k:-KZNZeiXtRVODgff0z_I:b-vcupk8) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L90)):*
> ```C
> int public_key_cmp(const uint8_t *pk1, const uint8_t *pk2)
> {
> #if crypto_box_PUBLICKEYBYTES != 32
> ```

Why here not `CRYPTO_PUBLIC_KEY_SIZE`?

---

*[toxcore/crypto_core.c, line 118 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZqPxv4emVmLIFTjM:-KZNZqPxv4emVmLIFTjN:b-tjtcxh) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L118)):*
> ```C
>  * return 1 if it is.
>  */
> bool public_key_valid(const uint8_t *public_key)
> ```

Update docs, please

---

*[toxcore/crypto_core.h, line 92 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaNFcq0mL9FNUz6d7:-KZNaNFdMUnmJtWe7cDc:bhw2wfl) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L92)):*
> ```objc
>  * timing attacks could reveal that data.
>  */
> int32_t crypto_memcmp(const void *p1, const void *p2, size_t length);
> ```

Are you sure, that time of memcpy independent of input data?

---

*[toxcore/crypto_core.h, line 100 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaU1q8frYHQVoEfx2:-KZNaU1q8frYHQVoEfx3:b1i4ash) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L100)):*
> ```objc
>  * memory is indeed zeroed.
>  */
> void crypto_memzero(void *data, size_t length);
> ```

But inside you use `memset`. Is it right?

---

*[toxcore/crypto_core.h, line 140 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNcl7cEK4jh4HfEO5X:-KZNcl7dQpyCNl2db4PA:biz36vm) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L140)):*
I think not return, but fill.

@return 0
> The crypto_box_keypair function randomly generates a secret key and a corresponding public key. It puts the secret key into sk[0], sk[1], ..., sk[crypto_box_SECRETKEYBYTES-1] and puts the public key into pk[0], pk[1], ..., pk[crypto_box_PUBLICKEYBYTES-1]. It then returns 0. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 2 of 47 files reviewed at latest revision, 7 unresolved discussions.

---

*[toxcore/crypto_core.c, line 48 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZxLmyJE974Z508nB:-KZNgUHbRkUMjgsidfQG:bxc3rkn) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L48)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

What do you know?)
</blockquote></details>

I know that it's insecure. Moved the comment down and added some more words.

---

*[toxcore/crypto_core.c, line 88 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNbB1NaPmQjP7-bAOF:-KZNgySnX9gMq039wqC5:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L88)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

In `.h` file you specified `int32_t`. Same for all other `int` occupation
</blockquote></details>

Done.

---

*[toxcore/crypto_core.c, line 90 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZeiW6tZKWR0Enn3k:-KZNgzGdFC4-LV8ImFL2:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L90)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Why here not `CRYPTO_PUBLIC_KEY_SIZE`?
</blockquote></details>

Done.

---

*[toxcore/crypto_core.c, line 118 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZqPxv4emVmLIFTjM:-KZNh-GTqAvvW7x3oPBm:bvu8yjw) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L118)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Update docs, please
</blockquote></details>

Removed docs. They are in the .h file already.

---

*[toxcore/crypto_core.h, line 92 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaNFcq0mL9FNUz6d7:-KZNh2jQ-dC2oG6vi6T6:b-pgaz2s) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L92)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Are you sure, that time of memcpy independent of input data?
</blockquote></details>

Yes. Not independent of input length, but any N sized string should take the same time. This means for instance that comparing "aaaa" and "aaaa" takes 4 time, and "aaaa" and "baaa" also takes 4 time. With a regular memcmp, the latter may take 1 time, because it immediately knows that the two strings are not equal.

---

*[toxcore/crypto_core.h, line 100 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaU1q8frYHQVoEfx2:-KZNhPU6Bj8gDFATnfIM:b-lw6sq1) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L100)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

But inside you use `memset`. Is it right?
</blockquote></details>

Yes, that is not good, but fixing it is not in scope for this PR.

---

*[toxcore/crypto_core.h, line 140 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNcl7cEK4jh4HfEO5X:-KZNhXy67P8IoGGQbg3e:bhln7yo) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L140)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

I think not return, but fill.

@return 0
> The crypto_box_keypair function randomly generates a secret key and a corresponding public key. It puts the secret key into sk[0], sk[1], ..., sk[crypto_box_SECRETKEYBYTES-1] and puts the public key into pk[0], pk[1], ..., pk[crypto_box_PUBLICKEYBYTES-1]. It then returns 0. 
</blockquote></details>

Correct. Fixed docs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Reviewed 43 of 45 files at r1, 2 of 3 files at r2.
Review status: 43 of 47 files reviewed at latest revision, 11 unresolved discussions.

---

*[auto_tests/crypto_test.c, line 148 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNejQIhx7rCgzuGFT6:-KZNejQJwc6CLzEIBoq8:b-kkb5hz) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/crypto_test.c#L148)):*
> ```C
>     unsigned char n[CRYPTO_NONCE_SIZE];
> 
>     unsigned char m[500];
> ```

What is 500?

---

*[auto_tests/dht_test.c, line 170 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNfoT3iSqM86Okf6Tr:-KZNfoT49_FQ-fOgS4tk:bolg1b1) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/dht_test.c#L170)):*
> ```C
>     // check "bad" clients replacement
>     int used, test1, test2, test3;
>     uint8_t public_key[CRYPTO_PUBLIC_KEY_BYTES], test_id1[CRYPTO_PUBLIC_KEY_SIZE], test_id2[CRYPTO_PUBLIC_KEY_SIZE],
> ```

`s/CRYPTO_PUBLIC_KEY_BYTES/CRYPTO_PUBLIC_KEY_SIZE/`

---

*[auto_tests/dht_test.c, line 214 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNhFHZVpIYuPK3gBgU:-KZNhFHZVpIYuPK3gBgV:bolg1b1) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/dht_test.c#L214)):*
> ```C
>     // check "possibly bad" clients replacement
>     int used, test1, test2, test3;
>     uint8_t public_key[CRYPTO_PUBLIC_KEY_BYTES], test_id1[CRYPTO_PUBLIC_KEY_SIZE], test_id2[CRYPTO_PUBLIC_KEY_SIZE],
> ```

`s/CRYPTO_PUBLIC_KEY_BYTES/CRYPTO_PUBLIC_KEY_SIZE/`

---

*[toxencryptsave/toxencryptsave.c, line 40 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNkhUIcoqhOzAExF_9:-KZNkhUIcoqhOzAExF_A:b-mqqtmf) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxencryptsave/toxencryptsave.c#L40)):*
> ```C
> #endif
> 
> #include <string.h>
> ```

Why do you add this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 43 of 47 files reviewed at latest revision, 8 unresolved discussions.

---

*[auto_tests/crypto_test.c, line 148 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNejQIhx7rCgzuGFT6:-KZNm14aOm4fdunw_rwU:blt7ctc) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/crypto_test.c#L148)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

What is 500?
</blockquote></details>

Some number of bytes. I didn't write this, and I'm not changing it in this PR.

---

*[auto_tests/dht_test.c, line 170 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNfoT3iSqM86Okf6Tr:-KZNm73-nM9eLzDyNXky:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/dht_test.c#L170)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`s/CRYPTO_PUBLIC_KEY_BYTES/CRYPTO_PUBLIC_KEY_SIZE/`
</blockquote></details>

Done.

---

*[auto_tests/dht_test.c, line 214 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNhFHZVpIYuPK3gBgU:-KZNm7ZMvVGcdE_hhBvM:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/auto_tests/dht_test.c#L214)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`s/CRYPTO_PUBLIC_KEY_BYTES/CRYPTO_PUBLIC_KEY_SIZE/`
</blockquote></details>

Done.

---

*[toxencryptsave/toxencryptsave.c, line 40 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNkhUIcoqhOzAExF_9:-KZNm8GcoKzRFGrS8pBC:bw31xof) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxencryptsave/toxencryptsave.c#L40)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Why do you add this?
</blockquote></details>

For `memset`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/crypto_core.c, line 118 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZqPxv4emVmLIFTjM:-KZNl_h5gfoL_19uE3XX:bvvqphy) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L118)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Removed docs. They are in the .h file already.
</blockquote></details>

Seems, you forgot to remove it

---

*[toxcore/crypto_core.c, line 282 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNmMJjOZSpdQMx24tQ:-KZNmMJkTdJ7xf_arYUD:b-qy2bj8) ([raw file](https://github.com/toktok/c-toxcore/blob/0b089172a35e7903caa2e4b0b1a2e19d1770cbf7/toxcore/crypto_core.c#L282)):*
> ```C
> {
> #ifdef VANILLA_NACL
>     /* I know: this is insecure. We need to provide our own secure
> ```

I think, adding `TODO` will be helpful

---

*[toxcore/crypto_core.c, line 293 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNmWQcNAJv9GvyH2sv:-KZNmWQcNAJv9GvyH2sw:b-ya81ng) ([raw file](https://github.com/toktok/c-toxcore/blob/0b089172a35e7903caa2e4b0b1a2e19d1770cbf7/toxcore/crypto_core.c#L293)):*
> ```C
> {
> #ifdef VANILLA_NACL
>     /* I know */
> ```

Same as above

---

*[toxcore/crypto_core.h, line 100 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaU1q8frYHQVoEfx2:-KZNm-MYmOLjvrb4luSF:b-3hp9i8) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L100)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Yes, that is not good, but fixing it is not in scope for this PR.
</blockquote></details>

Hm... It's added in this PR. You want to merge some vulnerable code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 45 of 47 files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/crypto_core.c, line 118 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNZqPxv4emVmLIFTjM:-KZNnUWuNDN8RJ6HXZks:b-k3o6xq) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.c#L118)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Seems, you forgot to remove it
</blockquote></details>

Right, I removed a different comment :). Done.

---

*[toxcore/crypto_core.c, line 282 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNmMJjOZSpdQMx24tQ:-KZNo3ShG4k8vSRjJWwv:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/0b089172a35e7903caa2e4b0b1a2e19d1770cbf7/toxcore/crypto_core.c#L282)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

I think, adding `TODO` will be helpful
</blockquote></details>

Done.

---

*[toxcore/crypto_core.c, line 293 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNmWQcNAJv9GvyH2sv:-KZNoJTwBbqfaYETGehz:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/0b089172a35e7903caa2e4b0b1a2e19d1770cbf7/toxcore/crypto_core.c#L293)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.

---

*[toxcore/crypto_core.h, line 100 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZNaU1q8frYHQVoEfx2:-KZNoKbottITI_5FRupH:b-nbinmo) ([raw file](https://github.com/toktok/c-toxcore/blob/48bc81f8d7f91b26629d875ac8fdac7a240e12fc/toxcore/crypto_core.h#L100)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Hm... It's added in this PR. You want to merge some vulnerable code?
</blockquote></details>

No, it's the same. It was like this before.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-:-KZNoecbFbyKfnAdRodA:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 2 files at r4.
Review status: 46 of 47 files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/crypto_core.h, line 45 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOcakm-0kuC0Cm1Qgs:-KZOcakm-0kuC0Cm1Qgt:b-yd7ec) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L45)):*
> ```objc
> 
> /**
>  * The number of bytes in a shared key computed from public and secret key.
> ```

`s/key/keys/`

---

*[toxcore/crypto_core.h, line 52 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOcfPt31doGE4kBeze:-KZOcfPt31doGE4kBezf:b-nbfg2n) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L52)):*
> ```objc
> 
> /**
>  * The number of bytes in a random symmetric key.
> ```

maybe `s/random //` ?

---

*[toxcore/crypto_core.h, line 120 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOdg4Gi88R3EPId20W:-KZOdg4Gi88R3EPId20X:by37bce) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L120)):*
>@return 0 if both mem locations of length are equal, -1 if they are not.

What is "mem location of length"? Probably should say "0 if both keys are equal, -1 if they are not".

Might also mane sense to return a `bool`, though might be a bit too much work to edit all the code.

---

*[toxcore/crypto_core.h, line 157 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOjMINUDxPniuLHTIO:-KZOjMIO4V2XULC85Idx:b-cv2lql) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L157)):*
> ```objc
>  * CRYPTO_MAC_SIZE using the public key (CRYPTO_PUBLIC_KEY_SIZE bytes) of the
>  * receiver and the secret key of the sender and a CRYPTO_NONCE_SIZE byte
>  * nonce.
> ```

Uh, I stopping commenting on the comments. Way too many comments in this file need to be improved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 1 files at r5.
Review status: 46 of 47 files reviewed at latest revision, 8 unresolved discussions.

---

*[toxcore/crypto_core.c, line 32 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOolDu0R5UTMK-LXbK:-KZOolDvF4tNTUGJ9iiX:bpm8e1c) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L32)):*
> ```C
> #include "crypto_core.h"
> 
> #include "network.h"
> ```

Network? Really? What does it use from `network.h`?

---

*[toxcore/crypto_core.c, line 90 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOk_j98f1bF8FtMbip:-KZOk_j98f1bF8FtMbiq:b60i5po) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L90)):*
> ```C
> }
> 
> uint32_t random_int(void)
> ```

Shouldn't it be static?

---

*[toxcore/crypto_core.c, line 97 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOkes0RhuqyHMzv0wp:-KZOkes0RhuqyHMzv0wq:b60i5po) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L97)):*
> ```C
> }
> 
> uint64_t random_64b(void)
> ```

Shouldn't it be static?

---

*[toxcore/crypto_core.c, line 118 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOkvm8VTsIcsNFVyhx:-KZOkvm8VTsIcsNFVyhy:bv6tztu) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L118)):*
> ```C
>  * enc_key has to be crypto_box_BEFORENMBYTES bytes long.
>  */
> int32_t encrypt_precompute(const uint8_t *public_key, const uint8_t *secret_key, uint8_t *enc_key)
> ```

`s/enc_key/shared_key/`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 46 of 47 files reviewed at latest revision, 8 unresolved discussions.

---

*[toxcore/crypto_core.c, line 32 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOolDu0R5UTMK-LXbK:-KZOufwZPPC8a3gj56is:bayu27f) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L32)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Network? Really? What does it use from `network.h`?
</blockquote></details>

`htonl`. I didn't want to replicate the "if win32 then include this with these defines etc" logic. It'll go away when we stop using `htonl`. I guess (without having seen it) the function that uses it can easily be rewritten to just not use `htonl`, but that's not in scope for this PR.

---

*[toxcore/crypto_core.c, line 90 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOk_j98f1bF8FtMbip:-KZOv0W3D9bNGN1Cz_VX:b-itdi0r) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L90)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Shouldn't it be static?
</blockquote></details>

No, it's part of the crypto_core public API. Making it static would make it internal linkage, so other files can't use it. It is static in the api.h.

---

*[toxcore/crypto_core.c, line 97 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOkes0RhuqyHMzv0wp:-KZOv8UGXLK5taRqOpKK:b-3rdboq) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L97)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Shouldn't it be static?
</blockquote></details>

No. See above.

---

*[toxcore/crypto_core.c, line 118 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOkvm8VTsIcsNFVyhx:-KZOvFwUq_nxbOkZekyD:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.c#L118)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

`s/enc_key/shared_key/`
</blockquote></details>

Done.

---

*[toxcore/crypto_core.h, line 45 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOcakm-0kuC0Cm1Qgs:-KZOvLdHuRP9szil42Vm:bfw5gs0) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L45)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

`s/key/keys/`
</blockquote></details>

Debatable, but done.

---

*[toxcore/crypto_core.h, line 52 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOcfPt31doGE4kBeze:-KZOvPvyRt541SgPE-w8:b-wxi5em) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L52)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

maybe `s/random //` ?
</blockquote></details>

Yeah, although this whole constant should probably go away. It's always used in the context of shared keys, so we should just use shared key size. That will be done after this PR.

---

*[toxcore/crypto_core.h, line 120 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOdg4Gi88R3EPId20W:-KZOvYi_z88ph8c8pBkC:b2yomnb) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L120)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

>@return 0 if both mem locations of length are equal, -1 if they are not.

What is "mem location of length"? Probably should say "0 if both keys are equal, -1 if they are not".

Might also mane sense to return a `bool`, though might be a bit too much work to edit all the code.
</blockquote></details>

Improving docs will happen later.

---

*[toxcore/crypto_core.h, line 157 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZOjMINUDxPniuLHTIO:-KZOvePMOpVJhISGj8ig:bwk9jsc) ([raw file](https://github.com/toktok/c-toxcore/blob/6b5e9ab12e408f79440674a59512e97c324a90af/toxcore/crypto_core.h#L157)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Uh, I stopping commenting on the comments. Way too many comments in this file need to be improved.
</blockquote></details>

Right. Later :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Reviewed 1 of 1 files at r6, 3 of 3 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[toxcore/crypto_core.c, line 176 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZPupmMwLmsvxCcgGeX:-KZPupmMwLmsvxCcgGeY:b-6uxegk) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/toxcore/crypto_core.c#L176)):*
> ```C
>     encrypt_precompute(public_key, secret_key, k);
>     int ret = encrypt_data_symmetric(k, nonce, plain, length, encrypted);
>     sodium_memzero(k, sizeof k);
> ```

Maybe change to `ctypto_memzero`?
Same for other occupations and for `sodium_memcmp`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

robinlinden commented:
Reviewed 8 of 45 files at r1, 1 of 1 files at r6, 1 of 3 files at r7, 1 of 2 files at r8.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[auto_tests/messenger_test.c, line 28 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZQRQ2A2C7qYX0ced_O:-KZQRQ2BOiPsuwRvUC4R:b-nxz2zh) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/auto_tests/messenger_test.c#L28)):*
> ```C
> 
> #if VANILLA_NACL
> #include <crypto_box.h> // crypto_box_PUBLICKEYBYTES and other defines.
> ```

Are we still going to use these defines?

---

*[auto_tests/messenger_test.c, line 140 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZQQ-iPWfPfLgU6tpgi:-KZQQ-iQuysNQDirHgii:bu9m7mg) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/auto_tests/messenger_test.c#L140)):*
> ```C
>     int good_len = strlen(good_data);
>     int bad_len = strlen(bad_data);
>     int really_bad_len = (MAX_CRYPTO_PACKET_SIZE - crypto_box_PUBLICKEYBYTES
> ```

Still using `crypto_box_PUBLICKEYBYTES` etc. here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 47 of 48 files reviewed at latest revision, 11 unresolved discussions.

---

*[auto_tests/messenger_test.c, line 28 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZQRQ2A2C7qYX0ced_O:-KZRV5hW3eRHJBlt2Yq3:b-4zhr3k) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/auto_tests/messenger_test.c#L28)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Are we still going to use these defines?
</blockquote></details>

Yes, I added these includes for them.

---

*[auto_tests/messenger_test.c, line 140 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZQQ-iPWfPfLgU6tpgi:-KZRV7ym_e16BgxJ4s-R:b-itxjuf) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/auto_tests/messenger_test.c#L140)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Still using `crypto_box_PUBLICKEYBYTES` etc. here.
</blockquote></details>

Yes, although I'd prefer to use our own copies of the constants, this expression involves BOXZEROBYTES and ZEROBYTES, which we intentionally don't expose. I don't know what the expression does (I don't want to know right now), but obviously it knows details about libsodium/nacl. I'll let it continue knowing that for now.

---

*[toxcore/crypto_core.c, line 176 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-KZPupmMwLmsvxCcgGeX:-KZRVawTsg79Fx5iSFY1:b-280uej) ([raw file](https://github.com/toktok/c-toxcore/blob/2eb2fe01a45c4603b6f6360551ca541c9c18a18a/toxcore/crypto_core.c#L176)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Maybe change to `ctypto_memzero`?
Same for other occupations and for `sodium_memcmp`
</blockquote></details>

Done. I'm leaving toxencryptsave mostly alone because it already needs to know a lot more about libsodium than crypto_core exposes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343)*
<!-- Sent from Reviewable.io -->

robinlinden commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 36 of 45 files at r1, 2 of 3 files at r2, 1 of 3 files at r7, 1 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-:-KZSjWQxAGNezWcwoqhp:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Now we need to improve the comments in crypto_core.h.

It seems that the `new_symmetric_key()` function gets misused a lot in toxcore and it usage needs to be replaced with`random_bytes()` in many places. Some of those places are its usage in `tox.c`, `group.c` and `onion_announce.c`. I mean, just look at this

```C
toxcore/group.h:#define GROUP_IDENTIFIER_LENGTH (1 + CRYPTO_SYMMETRIC_KEY_SIZE) /* type + CRYPTO_SYMMETRIC_KEY_SIZE so we can use new_symmetric_key(...) to fill it */
toxcore/onion_announce.h:    /* This is CRYPTO_SYMMETRIC_KEY_SIZE long just so we can use new_symmetric_key() to fill it */
```


<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 38 of 45 files at r1, 2 of 3 files at r2, 1 of 1 files at r6, 2 of 3 files at r7, 2 of 2 files at r8, 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-:-KZ_GsDPRVVQBlxAu7LS:b6uqg39)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Just to make it clear, the things I have described in my last comment are for the next PR. This PR looks good to get merged.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/343#-:-KZ_MjswNYpfWQvq--vm:b-rb88xf)*
<!-- Sent from Reviewable.io -->
